### PR TITLE
rh-che #504: Switching to OSO-proxy for all communications instead of calling OSO directly

### PR DESCRIPTION
**Pull request by ibuziuk**
### What does this PR do?
Switching to OSO-proxy for all communications instead of calling OSO directly

### What issues does this PR fix or reference?
https://github.com/redhat-developer/rh-che/issues/504

### How have you tested this PR?
Not tested yet, need to add test che server route to valid `redirect url` list on prod-preview

### CODE REVIEW

**Comment by l0rd** [comment]
@ibuziuk I see that you are using https://f8osoproxy-test-dsaas-preview.b6ff.rh-idev.openshiftapps.com as `che-fabric8-multicluster-oso-proxy-url`. So what is the plan? Are we going to promote it to PROD with this test URL or do you plan to open a second PR later?

**Comment by ibuziuk** [comment]
@l0rd cm here is just a placeholder and we would need to update cms on prod / prod-preview separately. Also, it looks like that no changes are required upstream (probably I would just deprecate obtaining oso token functionality) so the plan is do multi-cluster support in one (this) PR. So my plan is:
- test it locally on minishift agaist prod-preview (waiting for @alexeykazakov  to grant me access to prod-preview keycloak)
- if all works fine, ask SD team to add `che-fabric8-multicluster-oso-proxy-url` to prod-preview `che` cm and merge this PR for testing on prod-preview

Once oso-proxy will be available on prod (currently it is only on prod-preview) update prod `che` cm & promote the fix to prod. 
Does it sound like a plan?

**Comment by l0rd** [comment]
@ibuziuk 👍

**Comment by ibuziuk** [comment]
tested on dev.rdu2c.fabric8.io and oso proxy seems to work just fine. We could proceed with merging next week 

image - ibuziuk/che-server-multiuser:c82f44b-fabric8-4d75775
che-server - https://che-multi-cluster-fix.dev.rdu2c.fabric8.io/dashboard/

**Comment by l0rd** [comment]
@ibuziuk I have merge multi-tenant-rh-che to master branch so you should change the target of this PR now

**Comment by ibuziuk** [comment]
@alexeykazakov @aslakknutsen @l0rd  after discussing on daily it was decided not to merge / promote this PR to prod-preview because until prod does not support OSO proxy we will be blocked with promoting new che / rh-che fixes to master (revert of this fix before promoting to master would be required). Basically, there are 2 options:

A. Wait until oso-proxy will be available on master. In the meantime deploy che with oso-proxy support on dev cluster (dev.rdu2c) for testing

B. Change this PR so that 2 versions of OSO API would be supported (proxy / non-proxy) e.g. if `che-fabric8-multicluster-oso-proxy-url` is set and not blank -  use it for OSO communications, if not - use the old version with token lookup.

WDYT ?

**Comment by rhopp** [comment]
Do we have ETA for A. ?

**Comment by alexeykazakov** [comment]
@ibuziuk I agree that's you will block Che deployment to prod if you merge changes which will fail in production (because of missing dependency - proxy) to master. But A or B it's up to you :) Whatever is easier for you (to hold this PR and take care of possible conflicts with other PRs if any or support both singe-cluster and multi-cluster for now if it's a simple task).

**Comment by ibuziuk** [comment]
@alexeykazakov @alexeykazakov ok, I will stick to plan A which is easier, less error prone and without some config overhead. Please, let me know once proxy would be on prod. In the mean time che version with oso-proxy support is deployed to dev cluster:

- https://che-new-multicluster-fix.dev.rdu2c.fabric8.io/dashboard/
image: `ibuziuk/che-server-multiuser:e06ac73-fabric8-4031bbe`

**Comment by aslakknutsen** [comment]
@ibuziuk The proxy is in prod. Can be reached at https://f8osoproxy-test-dsaas-production.09b5.dsaas.openshiftapps.com for now. Fully functional as is, but will get a new name at some point. But that's just a config change so code can be deployed

**Comment by ibuziuk** [comment]
gitlab issue for cm update on prod / prod-preview https://gitlab.cee.redhat.com/dtsd/housekeeping/issues/1160

File: plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/multicluster/ClusterToRoutingSuffixMapping.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * @see UserBasedWorkspacesRoutingSuffixProvider
+ * @author Ilya Buziuk <ibuziuk@redhat.com>
+ */
+public final class ClusterToRoutingSuffixMapping {
+
+  private static final Map<String, String> CLUSTER_ROUTING_SUFFIX_MAPPING =
+      ImmutableMap.<String, String>builder()
+          .put("https://api.free-stg.openshift.com", "1b7d.free-stg.openshiftapps.com")
+          .put(
+              "https://api.starter-us-east-2.openshift.com",
```
Comment (davidfestal) [inline plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/multicluster/ClusterToRoutingSuffixMapping.java]: So for now there's only 1 cluster per environment (prod and prod-preview) ?

Comment (ibuziuk) [reply plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/multicluster/ClusterToRoutingSuffixMapping.java]: correct. BTW, This class is just workaround till https://github.com/fabric8-services/fabric8-wit/issues/1859 is fixed

Comment (davidfestal) [reply plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/multicluster/ClusterToRoutingSuffixMapping.java]: OK, thanks !

**Pull request merged**